### PR TITLE
perf: improve derp connection establishment

### PR DESCRIPTION
**Pull request by dignifiedquire**
- make sure that TLS session resumption is possible, by caching the TLS config in the derper
- parallelise dials to derpers in the same region
- expand the `doctor derp-regions` command

### CODE REVIEW

File: iroh-net/src/derp/http/client.rs [hunk 1]
```diff
@@ -620,28 +645,49 @@ impl Client {
...
-            let conn = self.dial_node(node).await;
-            match conn {
-                Ok(conn) => return Ok((conn, node.clone())),
-                Err(e) => first_err = Some(e),
             }
         }
-        let err = first_err.unwrap();
-        Err(err)
+
+        Err(first_err.unwrap())
```
Comment (rklaehn) [inline iroh-net/src/derp/http/client.rs:680]: seems a bit arbitrary to just return the first error.

**Review by rklaehn — APPROVED** [review]

File: iroh-net/src/derp/http/client.rs [hunk 2]
```diff
@@ -624,24 +640,44 @@ impl Client {
...
+                let target = target.clone();
+                async move {
+                    if node.stun_only {
+                        return Err(ClientError::StunOnlyNodesFound(target));
+                    }
+                    let conn = this.dial_node(&node).await;
+                    match conn {
+                        Ok(conn) => Ok((conn, node)),
+                        Err(e) => Err(e),
+                    }
```
Comment (flub) [inline iroh-net/src/derp/http/client.rs:659]: optional, but isn't this `this.dial_node(&node).await.map(|c| (c, node))`?

File: iroh-net/src/derp/http/client.rs [hunk 3]
```diff
@@ -624,24 +640,44 @@ impl Client {
...
+        while let Some(res) = dials.next().await {
+            match res {
+                Ok((conn, node)) => {
+                    // return on the first successfull one
+                    trace!("dialed region");
+                    return Ok((conn, node));
+                }
+                Err(e) => {
+                    if first_err.is_none() {
+                        first_err = Some(e);
```
Comment (flub) [inline iroh-net/src/derp/http/client.rs:674]: related to @rklaehn 's comment below, unless you return a Vec of errors you'll always have an arbitrary choice.  Probably easier is to `tracing::error!()` log all the error cases here?  or at least `warn!`.

**Review by flub — APPROVED** [review]

File: iroh-net/src/derp/http/client.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh-net/src/derp/http/client.rs]: fixed

File: iroh-net/src/derp/http/client.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-net/src/derp/http/client.rs]: fixed

File: iroh-net/src/derp/http/client.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-net/src/derp/http/client.rs]: logging them at least now

**Pull request merged**